test/TEST-64-UDEV-STORAGE/test.sh

```diff
@@ -0,0 +1,193 @@
+        "-device megasas-gen2,id=scsi3"
+    )
+
+    for i in {0..127}; do
+        # Add 128 drives, 32 per bus
+        qemu_opts+=(
+            "-device scsi-hd,drive=drive$i,bus=scsi$((i / 32)).0,channel=0,scsi-id=$((i % 32)),lun=0"
+            "-drive format=raw,cache=unsafe,file=${TESTDIR:?}/disk$i.img,if=none,id=drive$i"
+        )
+    done
+
+    KERNEL_APPEND="systemd.setenv=TEST_FUNCTION_NAME=${FUNCNAME[0]} ${USER_KERNEL_APPEND:-}"
+    QEMU_OPTIONS="${qemu_opts[*]} ${USER_QEMU_OPTIONS:-}"
+    test_run_one "${1:?}"
+}
+
+testcase_nvme_basic() {
+    if ! "${QEMU_BIN:?}" -device help | grep 'name "nvme"'; then
+        echo "nvme device driver is not available, skipping test..."
+        return 77
+    fi
+
+    for i in {0..27}; do
+        qemu_opts+=(
+            "-device nvme,drive=nvme$i,serial=deadbeef$i,num_queues=8"
+            "-drive format=raw,cache=unsafe,file=${TESTDIR:?}/disk$i.img,if=none,id=nvme$i"
+        )
+    done
+
+    KERNEL_APPEND="systemd.setenv=TEST_FUNCTION_NAME=${FUNCNAME[0]} ${USER_KERNEL_APPEND:-}"
+    QEMU_OPTIONS="${qemu_opts[*]} ${USER_QEMU_OPTIONS:-}"
+    test_run_one "${1:?}"
+}
+
+# Test for issue https://github.com/systemd/systemd/issues/20212
+testcase_virtio_scsi_identically_named_partitions() {
+    if ! "${QEMU_BIN:?}" -device help | grep 'name "virtio-scsi-pci"'; then
+        echo "virtio-scsi-pci device driver is not available, skipping test..."
+        return 77
+    fi
+
+    # Create 16 disks, with 8 partitions per disk (all identically named)
+    # and attach them to a virtio-scsi controller
+    local qemu_opts=("-device virtio-scsi-pci,id=scsi0,num_queues=4")
+    local diskpath="${TESTDIR:?}/namedpart0.img"
+    local lodev
+
+    # Save some time (and storage life) during local testing
+    if [[ ! -e "$diskpath" ]]; then
+        dd if=/dev/zero of="$diskpath" bs=1M count=18
+        lodev="$(losetup --show -f -P "$diskpath")"
+        sfdisk "${lodev:?}" <<EOF
+label: gpt
+
+name="Hello world", size=2M
+name="Hello world", size=2M
+name="Hello world", size=2M
+name="Hello world", size=2M
+name="Hello world", size=2M
+name="Hello world", size=2M
+name="Hello world", size=2M
+name="Hello world", size=2M
+EOF
+        losetup -d "$lodev"
+    fi
+
+    for i in {0..15}; do
+        diskpath="${TESTDIR:?}/namedpart$i.img"
+        if [[ $i -gt 0 ]]; then
+            cp -uv "${TESTDIR:?}/namedpart0.img" "$diskpath"
+        fi
+
+        qemu_opts+=(
+            "-device scsi-hd,drive=drive$i,bus=scsi0.0,channel=0,scsi-id=0,lun=$i"
+            "-drive format=raw,cache=unsafe,file=$diskpath,if=none,id=drive$i"
+        )
+    done
+
+    KERNEL_APPEND="systemd.setenv=TEST_FUNCTION_NAME=${FUNCNAME[0]} ${USER_KERNEL_APPEND:-}"
+    # Limit the number of VCPUs, disable KVM, and set a timeout to make sure
+    # we trigger the issue
+    QEMU_OPTIONS="${qemu_opts[*]} -machine accel=tcg ${USER_QEMU_OPTIONS:-}"
+    QEMU_SMP=1 QEMU_KVM=no QEMU_TIMEOUT=180 test_run_one "${1:?}"
```

Comment:
I do not think it is a good idea to make the test slower... It's hard to detect (tiny) performance regression on CI in general.
And, hopefully, #20603 made significant performance improvement, as you already tested previously. So, I think it is ok to run the test with the default KVM, SMP, and TIMEOUT settings.

If the test is very slow without KVM, then maybe we should skip the test when the system does not support KVM.

Reply:
I guess running the test _only_ with KVM makes sense, will tweak it.

The "ultimate" goal of this was to have some sort of a *baseline* which we can use to determine if there's indeed a regression (i.e. if the test triggers the `QEMU_TIMEOUT=`), which requires the test to run in a comparable environment. So I guess we should still leave the `QEMU_SMP=1`  to have some sort of "stable" environment and tweak the `QEMU_TIMEOUT=` accordingly.